infra/cifuzz/filestore/gitlab/__init__.py

```diff
@@ -0,0 +1,134 @@
+import tempfile
+
+import http_utils
+import filestore
+
+# pylint:disable=no-self-use,unused-argument
+
+
+class GitlabFilestore(filestore.BaseFilestore):
+  """Implementation of BaseFilestore using Gitlab job artifacts. Relies on
+  using a cache to get latest job id oto get a type of artifacts.
+  having a PRIVATE-TOKEN supplied to access the Gitlab API as seems
+  the only way to download an artifact from another job"""
+
+  def __init__(self, config):
+    super().__init__(config)
+    self.downloaded = set()
+    self.artifacts_dir = self.config.platform_conf.artifacts_dir
+    self.cache_dir = self.config.platform_conf.cache_dir
+    self.download_dir = self.config.platform_conf.download_dir
+    self.api_url = self.config.platform_conf.api_url
+
+  def _copy_from_dir(self, src, name, reason):
+    dest_dir = os.path.join(self.config.workspace, self.artifacts_dir, reason,
+                            name)
+    logging.info('Uploading %s to artifacts to %s.', reason, dest_dir)
+    shutil.copytree(src, dest_dir)
+    # Saves current job id in gitlab cache.
+    job_id = self.config.platform_conf.current_job_id
+    cache_file_path = os.path.join(self.config.workspace, self.cache_dir,
+                                   reason)
+    with open(cache_file_path, 'w', encoding='ascii') as cache_handle:
+      cache_handle.write(job_id)
+
+  def upload_crashes(self, name, directory):
+    """Gitlab artifacts implementation of upload_crashes."""
+    self._copy_from_dir(directory, name, 'crashes')
+
+  def upload_corpus(self, name, directory, replace=False):
+    """Gitlab artifacts implementation of upload_corpus."""
+    self._copy_from_dir(directory, name, 'corpus')
+
+  def upload_build(self, name, directory):
+    """Gitlab artifacts implementation of upload_build."""
+    self._copy_from_dir(directory, name, 'build')
+
+  def upload_coverage(self, name, directory):
+    """Gitlab artifacts implementation of upload_coverage."""
+    self._copy_from_dir(directory, name, 'coverage')
+
+  def _get_job_id(self, proj_path, reason):
+    """Get a specific job id for the latest succesful pipeline
+    with the specific job name."""
+    # First try to get job id from the cache.
+    cache_file_path = os.path.join(self.config.workspace, self.cache_dir,
+                                   reason)
+    with open(cache_file_path, 'r', encoding='ascii') as cache_handle:
+      job_id = int(cache_handle.read())
+      logging.info('Latest job from cache with %s is %d.', reason, job_id)
+      return job_id
+
+    # Otherwise, get job id from api.
+    # We could avoid PRIVATE-TOKEN and use only JOB-TOKEN
+    # by looping over all job ids until we find a relevant artifacts archive
+    headers = {'PRIVATE-TOKEN': self.config.platform_conf.private_token}
+    jobs_url = self.api_url + proj_path + '/jobs?scope=success'
+    with tempfile.NamedTemporaryFile() as tmp_file:
+      if not http_utils.download_url(jobs_url, tmp_file.name, headers=headers):
```

Comment:
Don't think we should download this to a file, just do it in memory.

Reply:
Is there a util function to do this ? 
`download_url` requires a file name and `get_json_from_url` does not allow you to set headers...
Should I add the optional parameter `headers` to `http_utils.get_json_from_url` ?

Reply:
I would add functionality to get_json_from_url to use headers rather than download to a file.